packages/relay/src/services/ReportService.ts

```diff
@@ -138,6 +143,92 @@ export class ReportService {
 
         return await db.findMany('ReportHistory', condition)
     }
+
+    async voteOnReport(
+        id: string,
+        nullifier: string,
+        adjudicateValue: AdjudicateValue,
+        db: DB
+    ) {
+        const report = await this.fetchSingleReport(id, db)
+        if (!report) {
+            throw InvalidReportError
+        }
+
+        if (report.status != ReportStatus.VOTING) {
+            throw NotVotingReportError
+        }
+
+        // check if user voted or not
+        if (this.isVoted(nullifier, report)) {
+            throw UserAlreadyVotedError
+        }
+
+        const adjudicatorsNullifier = this.upsertAdjudicatorsNullifier(
+            nullifier,
+            adjudicateValue,
+            report
+        )
+        // default value is 0, but insert statement doesn't have this field
+        // if this field is undefined, assume no one has voted yet.
+        const adjudicateCount = report.adjudicateCount ?? 0
+
+        // update adjudicatorsNullifier && adjudicateCount
+        await db.update('ReportHistory', {
+            where: {
+                reportId: id,
+            },
+            update: {
+                adjudicatorsNullifier: adjudicatorsNullifier,
+                adjudicateCount: adjudicateCount + 1,
+            },
+        })
+    }
+
+    upsertAdjudicatorsNullifier(
+        nullifier: string,
+        adjudicateValue: AdjudicateValue,
+        report: ReportHistory
+    ): Adjudicator[] {
+        const adjudicator = [
```

Comment:
This might be a cleaner way.
```
const newAdjudicator = { nullifier, adjudicateValue, claimed: false };

return report.adjudicatorsNullifier ? [...report.adjudicatorsNullifier, newAdjudicator] : [newAdjudicator];
```